whelktool/scripts/cleanups/2023/12/fmt-210-restore-primary-contributors.groovy

```diff
@@ -0,0 +1,77 @@
+/**
+ * Follow-up on LXL-2512 that moved instance bound contributions from work to instance.
+ * As a side effect of that move some agents expected to remain in the work's PrimaryContribution instead ended up in instance.contribution.
+ *
+ * See FMT-210 for details.
+ */
+
+unhandled = getReportWriter('unhandled.tsv')
+addedPrimary = getReportWriter('added-primary.tsv')
+
+def roles = [
+        'https://id.kb.se/relator/engraver',
+        'https://id.kb.se/relator/lithographer',
+        'https://id.kb.se/relator/etcher',
+        'https://id.kb.se/relator/woodcutter',
+        'https://id.kb.se/relator/woodEngraver',
+        'https://id.kb.se/relator/electrotyper'
+]
+        .collect { ['@id': it] } as Set
+
+def scriptId = "https://libris.kb.se/sys/globalchanges/2023/05/lxl-2512-move-contribution-by-relator-domain/script.groovy"
+
+selectBySqlWhere("collection = 'bib' and deleted = false and data#>>'{@graph,1,contribution}' is not null") { bib ->
+    def id = bib.doc.shortId
+    def instance = bib.graph[1]
+    def work = instance.instanceOf
+
+    if (!work || work['@id']) return
+
+    def newestToOldestVersion
+    try {
+        newestToOldestVersion = bib.getVersions().reverse()
+    } catch (Exception e) {
+        println(e.getStackTrace())
+        return
+    }
+    def versionTouchedByScriptIdx = newestToOldestVersion.findIndexOf { it.data['@graph'][0]['generationProcess']?['@id'] == scriptId }
```

Comment:
Shouldn't this lookup be in the reverse order? "oldestToNewest"
The oldest version with generationProcess == scriptId is the one that was touched by the script.

Reply:
You are right, I'll fix that.